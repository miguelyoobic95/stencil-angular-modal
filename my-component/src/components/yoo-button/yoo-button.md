---
name: Button
category: Core Components
---

Buttons can be used with `<yoo-button>` HTML tag.

## Attributes

|Name|Type|Default|Description|
|---|---|---|---|
|`title`|string|   |Text to be displayed in the button|
|`disabled`|boolean| `false`|Enable/Disable the button|
|`isLoading`|boolean|`false`|Displays loader animation in button|
|`icon`|string|     |Icon to add to the button|


```yoo-button.html
    <yoo-button title="Unstyled yoo-button"></yoo-button>
    <br>
    <yoo-button title="Button Disabled" disabled="true" class="accent"></yoo-button>
    <br>
    <yoo-button title="Button with icon" icon="yo-fire" class="accent"></yoo-button>
    <br>
    <yoo-button title="Button loading" is-loading="true"  class="accent"></yoo-button>
    <br>
```

```yoo-button.css hidden 
     yoo-button {
        display: block;
        margin-bottom: 5px;
    }
```


## CSS

Multiple button types may be specified via the class attribute.

|Type|Class|Usage|
|---|---|--------|
|Accent|`accent`| Provides extra visual weight and identifies the accent action in a set of buttons|
|Stable|`stable`|stable buttons are used for stable actions|
|Link|`link`| Standard, trasnparent link button
|Link transparent|`link transparent`|Toolbar transparent link button
|Danger|`danger`|Danger buttons are used for critical actions which are usually irreversible|
|Success|`success`| Indicates a successful or positive action|
|Info|`info`|Indicates an action that will provide information to the user|
|Gradient Accent|`gradient-accent`|button gradient accent|
|Gradient Danger|`gradient-danger`|button gradient danger|
|Gradient Info|`gradient-info`|button gradient info|
|Gradient Warning|`gradient-warning`|button gradient warning|
|Gradient Success|`gradient-success`|button gradient success|
|Gradient Grey|`gradient-grey`|button gradient grey|
|Icon Only|`icon-only`|Styles icon for icon-only format|
|Small|`small`|Small button|
|Block|`block`|Block button|
|Round|`round`|Rounded button|
```yoo-button-types.html
  <div attr-layout="row">
    <div attr-flex>
        <yoo-button title="Button accent" class="accent"></yoo-button>
        <br/>
        <yoo-button title="Button stable" class="stable"></yoo-button>
        <br/>
        <yoo-button title="Button link" class="link"></yoo-button>
        <br/>
        <yoo-button title="Button link transparent " class="link-transparent"></yoo-button>
        <br/>
        <yoo-button title="Button danger" class="danger"></yoo-button>
        <br/>
        <yoo-button title="Button info" class="info"></yoo-button>
        <br/>
        <yoo-button title="Button gradient-accent" class="gradient-accent"></yoo-button>
        <br/>
        <yoo-button title="Button gradient-accent" class="gradient-danger"></yoo-button>
        <br/>
        <yoo-button title="Button gradient-accent" class="gradient-info"></yoo-button>
        <br/>
        <yoo-button title="Button gradient-accent" class="gradient-warning"></yoo-button>
        <br/>
        <yoo-button title="Button gradient-accent" class="gradient-success"></yoo-button>
        <br/>
        <yoo-button title="Button gradient-accent" class="gradient-grey"></yoo-button>
        <br/>
        <yoo-button icon="yo-more-v" class="icon-only accent"></yoo-button>
        </br>
        <yoo-button icon="yo-plus" class="icon-only success"></yoo-button>
        <br/>
    </div>
        <div attr-flex>
        <yoo-button title="Button accent" class="round accent"></yoo-button>
        <br/>
        <yoo-button title="Button stable" class="round stable"></yoo-button>
        <br/>
        <yoo-button title="Button link" class="round link"></yoo-button>
        <br/>
        <yoo-button title="Button link transparent " class="rounded link-transparent"></yoo-button>
        <br/>
        <yoo-button title="Button danger" class="round danger"></yoo-button>
        <br/>
        <yoo-button title="Button info" class="round info"></yoo-button>
        <br/>
        <yoo-button title="Button gradient-accent" class="round gradient-accent"></yoo-button>
        <br/>
        <yoo-button title="Button gradient-accent" class="round gradient-danger"></yoo-button>
        <br/>
        <yoo-button title="Button gradient-accent" class="round gradient-info"></yoo-button>
        <br/>
        <yoo-button title="Button gradient-accent" class="round gradient-warning"></yoo-button>
        <br/>
        <yoo-button title="Button gradient-accent" class="round gradient-success"></yoo-button>
        <br/>
        <yoo-button title="Button gradient-accent" class="round gradient-grey"></yoo-button>
        <br/>
        <yoo-button icon="yo-more-v" class="rounded icon-only accent"></yoo-button>
        </br>
        <yoo-button icon="yo-plus" class="rounded icon-only success"></yoo-button>
        <br/>
    </div>
</div>
```
```yoo-button-types.css hidden
     yoo-button {
        display: block;
        margin-bottom: 5px;
    }
```

### Size Variations: 

**Small:**

Add `class="small"` for a small button:

```yoo-button-size.html
    <yoo-button title="Normal Button" class="accent"></yoo-button>
    <br/>
    <yoo-button title="Small button" class="accent small"></yoo-button>
    <br/>   
```
```yoo-button-size.css hidden
     yoo-button {
        display: block;
        margin-bottom: 5px;
    }
```

**Block:**

Add `class="block"` for a block button:

```yoo-button-block.html
    <yoo-button title="Button Block" class="accent block"></yoo-button>
    <br/>
```