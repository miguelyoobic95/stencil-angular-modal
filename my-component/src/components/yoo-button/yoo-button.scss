@import "../../styles/_variables.scss";

:host {
    .container {
        display: inline-flex;
        justify-content: center;
        padding: 0.5rem 1rem;
        border-radius: 5px;
        max-width: 15rem;
        cursor: pointer;
        transition: all 0.3s;
        border: 1px solid cssvar(grey-20);
        &:hover {
            box-shadow: $button-shadow;
            transform: translateY(-2px);
            background: cssvar(grey-05);
        }
        &:active,
        &:focus {
            outline: 0 !important;
        }
        &.disabled {
            pointer-events: none;
            opacity: 0.4;
        }
        .icon {
            font-size: 1rem;
            padding-right: 1rem;
            padding-left: 0.5rem;
        }
        .value {
            font-size: 1rem;
            line-height: 1.5;
            text-align: center;
            img {
                width: 1rem;
            }
        }
    }
}

:host(.round) {
    .container {
        border-radius: 2rem;
    }
}

:host(.block) {
    .container {
        display: block;
        width: 100%;
        max-width: 100%;
    }
}

:host(.icon-only) {
    .container {
        width: 2.625rem;
        .icon {
            padding-left: 0rem;
            padding-right: 0rem;
        }
    }
}

:host(.small) {
    .container {
        padding: 0.375rem 1rem;
        .value {
            font-size: 0.75rem;
            line-height: 1.5;
            text-align: center;
        }
    }
}

:host(.accent) {
    .container {
        background: cssvar(accent);
        border: 1px solid cssvar(accent);
        &:hover {
            background: cssvar(accent-110);
        }
        &:active {
            background: cssvar(accent-120);
        }
        .value {
            color: cssvar(light);
            fill: cssvar(light);
        }
    }
}

:host(.stable) {
    .container {
        background: cssvar(grey-10);
        border: 1px solid cssvar(grey-20);
        &:hover {
            background: cssvar(grey-20);
        }
        &:active {
            background: cssvar(grey-40);
        }
        .value {
            color: cssvar(grey-100);
        }
    }
}

:host(.link) {
    .container {
        background: cssvar(transparent);
        border: 1px solid cssvar(transparent);
        &:hover {
            background: cssvar(accent-05);
        }
        &:active {
            background: cssvar(accent-10);
        }
        .value {
            color: cssvar(accent);
            fill: cssvar(accent);
        }
    }
}

:host(.link-transparent) {
    .container {
        background: cssvar(transparent);
        border: 1px solid cssvar(transparent);
        color: cssvar(grey-40);
        fill: cssvar(grey-40);
        &:hover {
            color: cssvar(accent-110);
            background: cssvar(transparent);
            box-shadow: none;
            transform: translateY(0px);
        }
        &:active {
            background: cssvar(transparent);
            color: cssvar(accent-120);
        }
    }
}

:host(.danger) {
    .container {
        background: cssvar(danger);
        border: 1px solid cssvar(danger);
        &:hover {
            background: cssvar(danger-110);
        }
        &:active {
            background: cssvar(danger-120);
        }
        .value {
            color: cssvar(light);
            fill: cssvar(light);
        }
    }
}

:host(.success) {
    .container {
        background: cssvar(success);
        border: 1px solid cssvar(success);
        &:hover {
            background: cssvar(success-110);
        }
        &:active {
            background: cssvar(success-120);
        }
        .value {
            color: cssvar(light);
            fill: cssvar(light);
        }
    }
}

:host(.info) {
    .container {
        background: cssvar(info);
        border: 1px solid cssvar(info);
        &:hover {
            background: cssvar(info-110);
        }
        &:active {
            background: cssvar(info-120);
        }
        .value {
            color: cssvar(light);
            fill: cssvar(light);
        }
    }
}

:host(.gradient-accent) {
    .container {
        border: 1px solid cssvar(accent);
        background: cssvar(gradient-accent); // &:hover {
        //     background: cssvar(gradient-accent-110);
        // }
        // &:active {
        //     background: cssvar(gradient-accent-120);
        // }
        .value {
            color: cssvar(light);
            fill: cssvar(light);
        }
    }
}

:host(.gradient-info) {
    .container {
        border: 1px solid cssvar(info);
        background: cssvar(gradient-info); // &:hover {
        //     background: cssvar(gradient-info-110);
        // }
        // &:active {
        //     background: cssvar(gradient-info-120);
        // }
        .value {
            color: cssvar(light);
            fill: cssvar(light);
        }
    }
}

:host(.gradient-grey) {
    .container {
        border: 1px solid cssvar(grey);
        background: cssvar(gradient-grey); // &:hover {
        //     background: cssvar(gradient-grey-110);
        // }
        // &:active {
        //     background: cssvar(gradient-grey-120);
        // }
        .value {
            color: cssvar(light);
            fill: cssvar(light);
        }
    }
}

:host(.gradient-danger) {
    .container {
        border: 1px solid cssvar(danger);
        background: cssvar(gradient-danger); // &:hover {
        //     background: cssvar(gradient-danger-110);
        // }
        // &:active {
        //     background: cssvar(gradient-danger-120);
        // }
        .value {
            color: cssvar(light);
            fill: cssvar(light);
        }
    }
}

:host(.gradient-warning) {
    .container {
        border: 1px solid cssvar(warning);
        background: cssvar(gradient-warning); // &:hover {
        //     background: cssvar(gradient-warning-110);
        // }
        // &:active {
        //     background: cssvar(gradient-warning-120);
        // }
        .value {
            color: cssvar(light);
            fill: cssvar(light);
        }
    }
}

:host(.gradient-success) {
    .container {
        border: 1px solid cssvar(success);
        background: cssvar(gradient-success); // &:hover {
        //     background: cssvar(gradient-success-110);
        // }
        // &:active {
        //     background: cssvar(gradient-success-120);
        // }
        .value {
            color: cssvar(light);
            fill: cssvar(light);
        }
    }
}