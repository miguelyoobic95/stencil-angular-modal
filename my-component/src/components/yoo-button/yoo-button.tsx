import { Component, Prop, Event, EventEmitter } from '@stencil/core';

@Component({
    tag: 'yoo-button',
    styleUrl: 'yoo-button.scss',
    scoped: true
})
export class YooButtonComponent {

    @Prop() title: string;
    @Prop() disabled : boolean = false;
    @Prop() isLoading: boolean;
    @Prop() icon: string;    

    @Event() onClick: EventEmitter<boolean>;

    click() {
        this.onClick.emit(true);
    }
    getDefaultContainer() {
        return (
            <button class="container" onClick={() => this.click()}>
                <div class="value">
                    {this.icon ? <span class="icon"><i class={this.icon}></i></span> : null}
                    {this.title}
                </div>
            </button>
        );
    }

    getLoadingContainer() {
        return (
            <div class="container" onClick={() => this.click()}>
                <span class="value"><img src="assets/loader/loading.svg" /></span>
            </div>
        );
    }

    getDisabledContainer() {
        return (
            <button class="container disabled" disabled>
                <div class="value">
                    {this.icon ? <span class="icon"><i class={this.icon}></i></span> : null}
                    {this.title}
                    <slot></slot>
                </div>
            </button>
        );
    }
    render() {
        return (
            (this.isLoading ?  this.getLoadingContainer() : (this.disabled ? this.getDisabledContainer() : this.getDefaultContainer() ))
        );
    }

}
