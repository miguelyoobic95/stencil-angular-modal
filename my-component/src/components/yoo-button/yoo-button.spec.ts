import { render, flush } from '@stencil/core/testing';
import { YooButtonComponent } from './yoo-button';
//import * as domtoimage from 'dom-to-image';

describe('YooButtonComponent', () => {
    it('should build', () => {
        expect(new YooButtonComponent()).toBeTruthy();
    });

    describe('Snapshots', () => {
        it ('Should have text content', async () => {
            let element = await render({
               components: [YooButtonComponent],
               html: '<yoo-button title="Button accent" class="accent"></yoo-button>'
            });
            expect(element).toMatchSnapshot();
        });

        it ('Should have text content and icon', async () => {
            let element = await render({
               components: [YooButtonComponent],
               html: '<yoo-button title="Button accent" icon="yo-fire"></yoo-button>'
            });
            expect(element).toMatchSnapshot();
        });

        it ('Should have icon only', async () => {
            let element = await render({
               components: [YooButtonComponent],
               html: '<yoo-button title="Button loading" isLoading="true" class="icon-only"></yoo-button>'
            });
            expect(element).toMatchSnapshot();
        });

        it ('Should have loading icon and no title', async () => {
            let element = await render({
               components: [YooButtonComponent],
               html: '<yoo-button title="Button loading" isLoading="true" class="accent"></yoo-button>'
            });
            expect(element).toMatchSnapshot();
        });
    });

    describe('rendering', () => {
        it('Should re-render when isLoading is updated', async () => {
            let element = await render({
                components: [YooButtonComponent],
                html: '<yoo-button title="Button loading" class="accent"></yoo-button>'
             });
             expect(element.textContent).toEqual('Button loading');
             expect(element.querySelector('img')).toBeNull();
             element.isLoading = true;
             flush(element);
             expect(element.textContent).toEqual('');
             expect(element.querySelector('img').src).toMatch('loading.svg');
        });
    });
    
    describe('events', () => {

        it('should trigger click event', async () => {
            //let clicked = jest.fn();
            let element = await render({
                components: [YooButtonComponent],
                html: '<yoo-button title="Button loading" class="accent"></yoo-button>'
             });
             element.onclick = jest.fn();
             await element.click();
             expect(element.onclick).toBeCalled();
             
        });

        it('should not trigger click event when disabled', async () => {
            let clicked = jest.fn();
            let element = await render({
                components: [YooButtonComponent],
                html: '<yoo-button title="Button loading" disabled="true" class="accent"></yoo-button>'
             });
             let button = element.querySelector('button');
             button.addEventListener('click', clicked);
             await element.click();
             expect(clicked).not.toBeCalled();                         
             expect(element.disabled).toBeTruthy();
        });

    });
});