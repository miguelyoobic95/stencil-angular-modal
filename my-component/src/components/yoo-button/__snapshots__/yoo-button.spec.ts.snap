// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YooButtonComponent Snapshots Should have icon only 1`] = `
<yoo-button
  class="icon-only hydrated"
  data-yoo-button-host=""
  isloading="true"
  title="Button loading"
>
  <!---->
  <button
    class="container"
    data-yoo-button=""
  >
    <div
      class="value"
      data-yoo-button=""
    >
      Button loading
    </div>
  </button>
</yoo-button>
`;

exports[`YooButtonComponent Snapshots Should have loading icon and no title 1`] = `
<yoo-button
  class="accent hydrated"
  data-yoo-button-host=""
  isloading="true"
  title="Button loading"
>
  <!---->
  <button
    class="container"
    data-yoo-button=""
  >
    <div
      class="value"
      data-yoo-button=""
    >
      Button loading
    </div>
  </button>
</yoo-button>
`;

exports[`YooButtonComponent Snapshots Should have text content 1`] = `
<yoo-button
  class="accent hydrated"
  data-yoo-button-host=""
  title="Button accent"
>
  <!---->
  <button
    class="container"
    data-yoo-button=""
  >
    <div
      class="value"
      data-yoo-button=""
    >
      Button accent
    </div>
  </button>
</yoo-button>
`;

exports[`YooButtonComponent Snapshots Should have text content and icon 1`] = `
<yoo-button
  class="hydrated"
  data-yoo-button-host=""
  icon="yo-fire"
  title="Button accent"
>
  <!---->
  <button
    class="container"
    data-yoo-button=""
  >
    <div
      class="value"
      data-yoo-button=""
    >
      <span
        class="icon"
        data-yoo-button=""
      >
        <i
          class="yo-fire"
          data-yoo-button=""
        />
      </span>
      Button accent
    </div>
  </button>
</yoo-button>
`;
